File: cmd/shen-derive-ts/core/match.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { sym, num, bool_, sList, type Sexpr } from "./sexpr.ts";
import {
  match,
  intVal,
  boolVal,
  stringVal,
  listVal,
  type Value,
  type MatchResult,
} from "./match.ts";

// --- Helpers ---

// `(cons head tail)` pattern builder.
function cons(head: Sexpr, tail: Sexpr): Sexpr {
  return sList(sym("cons"), head, tail);
}

// `[A B]` desugars to (cons A (cons B nil)).
function fixedList(...items: Sexpr[]): Sexpr {
  let acc: Sexpr = sym("nil");
  for (let i = items.length - 1; i >= 0; i--) {
    acc = cons(items[i]!, acc);
  }
  return acc;
}

function expectMatched(r: MatchResult): Map<string, Value> {
  if (r.kind !== "matched") {
    assert.fail(`expected matched, got ${r.kind}${r.kind === "error" ? ": " + r.msg : ""}`);
  }
  return r.bindings;
}

// --- Tests ported from match_test.go ---

test("wildcard `_` matches anything and binds nothing", () => {
  const r = match(sym("_"), intVal(42));
  const b = expectMatched(r);
  assert.equal(b.size, 0);
});

test("uppercase variable binds the value", () => {
  const r = match(sym("X"), intVal(42));
  const b = expectMatched(r);
  const x = b.get("X");
  assert.ok(x);
  assert.deepEqual(x, intVal(42));
});

test("`nil` matches the empty list", () => {
  const r = match(sym("nil"), listVal());
  const b = expectMatched(r);
  assert.equal(b.size, 0);
});

test("`nil` does NOT match a non-empty list", () => {
  const r = match(sym("nil"), listVal(intVal(1)));
  assert.equal(r.kind, "miss");
});

test("cons pattern `[X | Xs]` binds head and tail", () => {
  const pat = cons(sym("X"), sym("Xs"));
  const lst = listVal(intVal(1), intVal(2), intVal(3));
  const b = expectMatched(match(pat, lst));
  assert.deepEqual(b.get("X"), intVal(1));
  const xs = b.get("Xs");
  assert.ok(xs && xs.kind === "list");
  assert.equal(xs.elems.length, 2);
  assert.deepEqual(xs.elems[0], intVal(2));
  assert.deepEqual(xs.elems[1], intVal(3));
});

test("cons pattern does not match empty list", () => {
  const pat = cons(sym("X"), sym("Xs"));
  assert.equal(match(pat, listVal()).kind, "miss");
});

test("fixed-length list `[A B]` matches a 2-element list", () => {
  const pat = fixedList(sym("A"), sym("B"));
  const b = expectMatched(match(pat, listVal(stringVal("x"), stringVal("y"))));
  assert.deepEqual(b.get("A"), stringVal("x"));
  assert.deepEqual(b.get("B"), stringVal("y"));
});

test("fixed-length `[A B]` does not match 1-element list", () => {
  const pat = fixedList(sym("A"), sym("B"));
  assert.equal(match(pat, listVal(stringVal("x"))).kind, "miss");
});

test("fixed-length `[A B]` does not match 3-element list", () => {
  const pat = fixedList(sym("A"), sym("B"));
  const v = listVal(stringVal("x"), stringVal("y"), stringVal("z"));
  assert.equal(match(pat, v).kind, "miss");
});

test("nested cons pattern binds inner and outer variables", () => {
  // [[X Y] | Rest]
  const pat = cons(fixedList(sym("X"), sym("Y")), sym("Rest"));
  const inner = listVal(stringVal("a"), stringVal("b"));
  const outer = listVal(inner, intVal(99));
  const b = expectMatched(match(pat, outer));
  assert.deepEqual(b.get("X"), stringVal("a"));
  assert.deepEqual(b.get("Y"), stringVal("b"));
  const rest = b.get("Rest");
  assert.ok(rest && rest.kind === "list");
  assert.equal(rest.elems.length, 1);
});

test("nested cons pattern misses when inner element wrong shape", () => {
  const pat = cons(fixedList(sym("X"), sym("Y")), sym("Rest"));
  const bad = listVal(listVal(stringVal("a")), intVal(99));
  assert.equal(match(pat, bad).kind, "miss");
});

test("boolean literal `true` matches BoolVal(true)", () => {
  assert.equal(match(bool_(true), boolVal(true)).kind, "matched");
  assert.equal(match(bool_(true), boolVal(false)).kind, "miss");
});

test("integer literal `42` matches IntVal(42)", () => {
  assert.equal(match(num(42), intVal(42)).kind, "matched");
  assert.equal(match(num(42), intVal(41)).kind, "miss");
});

test("cons pattern against int is a structural miss (not error)", () => {
  const pat = cons(sym("X"), sym("Xs"));
  const r = match(pat, intVal(1));
  assert.equal(r.kind, "miss");
});

test("malformed pattern: lowercase non-special symbol is an error", () => {
  const r = match(sym("foo"), intVal(1));
  assert.equal(r.kind, "error");
});

test("malformed pattern: non-cons list head is an error", () => {
  // (bogus X Y) should be rejected as a list pattern, not a miss.
  const pat = sList(sym("bogus"), sym("X"), sym("Y"));
  const r = match(pat, listVal(intVal(1), intVal(2)));
  assert.equal(r.kind, "error");
});

File: shen-derive/core/match_test.go
package core

import "testing"

// parsePattern is a shorthand that fails the test on parse error.
func parsePattern(t *testing.T, src string) Sexpr {
	t.Helper()
	s, err := ParseSexpr(src)
	if err != nil {
		t.Fatalf("parse %q: %v", src, err)
	}
	return s
}

func TestMatch_Wildcard(t *testing.T) {
	b, ok, err := Match(parsePattern(t, "_"), IntVal(42))
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if !ok {
		t.Fatal("expected match")
	}
	if len(b) != 0 {
		t.Errorf("expected no bindings, got %v", b)
	}
}

func TestMatch_VariableBinds(t *testing.T) {
	b, ok, err := Match(parsePattern(t, "X"), IntVal(42))
	if err != nil || !ok {
		t.Fatalf("match failed: ok=%v err=%v", ok, err)
	}
	if v, has := b["X"]; !has || v.String() != "42" {
		t.Errorf("expected X=42, got %v", b)
	}
}

func TestMatch_EmptyList(t *testing.T) {
	// [] parses as the symbol `nil`.
	b, ok, err := Match(parsePattern(t, "[]"), ListVal(nil))
	if err != nil || !ok {
		t.Fatalf("match: ok=%v err=%v", ok, err)
	}
	if len(b) != 0 {
		t.Errorf("expected no bindings, got %v", b)
	}

	// Non-empty list should NOT match `[]`.
	_, ok, err = Match(parsePattern(t, "[]"), ListVal{IntVal(1)})
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if ok {
		t.Error("expected mismatch for non-empty list against []")
	}
}

func TestMatch_ConsHeadTail(t *testing.T) {
	// [X | Xs] matches a non-empty list, binds head and tail.
	lst := ListVal{IntVal(1), IntVal(2), IntVal(3)}
	b, ok, err := Match(parsePattern(t, "[X | Xs]"), lst)
	if err != nil || !ok {
		t.Fatalf("match: ok=%v err=%v", ok, err)
	}
	if b["X"].String() != "1" {
		t.Errorf("X: got %v, want 1", b["X"])
	}
	xs, isList := b["Xs"].(ListVal)
	if !isList || len(xs) != 2 || xs[0].String() != "2" || xs[1].String() != "3" {
		t.Errorf("Xs: got %v, want [2 3]", b["Xs"])
	}

	// Empty list should NOT match `[X | Xs]`.
	_, ok, _ = Match(parsePattern(t, "[X | Xs]"), ListVal(nil))
	if ok {
		t.Error("empty list should not match cons pattern")
	}
}

func TestMatch_FixedLengthList(t *testing.T) {
	// [A B] desugars to (cons A (cons B nil)) — exactly a 2-element list.
	pat := parsePattern(t, "[A B]")

	b, ok, _ := Match(pat, ListVal{StringVal("x"), StringVal("y")})
	if !ok {
		t.Fatal("expected match for [x y]")
	}
	if b["A"].String() != "\"x\"" || b["B"].String() != "\"y\"" {
		t.Errorf("bindings: %v", b)
	}

	// 1-element list: should not match.
	_, ok, _ = Match(pat, ListVal{StringVal("x")})
	if ok {
		t.Error("1-element list matched [A B]")
	}

	// 3-element list: should not match (tail pattern `(cons B nil)` expects
	// a 1-element list, so matching against [y z] fails at nil).
	_, ok, _ = Match(pat, ListVal{StringVal("x"), StringVal("y"), StringVal("z")})
	if ok {
		t.Error("3-element list matched [A B]")
	}
}

func TestMatch_NestedPattern(t *testing.T) {
	// [[X Y] | Rest] — head is itself a 2-element list pattern.
	pat := parsePattern(t, "[[X Y] | Rest]")

	inner := ListVal{StringVal("a"), StringVal("b")}
	outer := ListVal{inner, IntVal(99)}
	b, ok, err := Match(pat, outer)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if !ok {
		t.Fatal("expected match")
	}
	if b["X"].String() != "\"a\"" || b["Y"].String() != "\"b\"" {
		t.Errorf("X/Y bindings: %v", b)
	}
	rest, isList := b["Rest"].(ListVal)
	if !isList || len(rest) != 1 {
		t.Errorf("Rest: got %v, want 1-element list", b["Rest"])
	}

	// If the head element isn't a 2-element list, the nested match fails.
	bad := ListVal{ListVal{StringVal("a")}, IntVal(99)}
	_, ok, _ = Match(pat, bad)
	if ok {
		t.Error("nested match should have failed: head is 1-element")
	}
}

func TestMatch_BooleanLiteral(t *testing.T) {
	pat := parsePattern(t, "true")
	_, ok, _ := Match(pat, BoolVal(true))
	if !ok {
		t.Error("true should match BoolVal(true)")
	}
	_, ok, _ = Match(pat, BoolVal(false))
	if ok {
		t.Error("true should not match BoolVal(false)")
	}
}

func TestMatch_IntLiteral(t *testing.T) {
	pat := parsePattern(t, "42")
	_, ok, _ := Match(pat, IntVal(42))
	if !ok {
		t.Error("42 should match IntVal(42)")
	}
	_, ok, _ = Match(pat, IntVal(41))
	if ok {
		t.Error("42 should not match IntVal(41)")
	}
}

func TestMatch_TypeMismatch(t *testing.T) {
	// Cons pattern against a non-list: structural miss, no error.
	_, ok, err := Match(parsePattern(t, "[X | Xs]"), IntVal(1))
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if ok {
		t.Error("cons pattern against int should not match")
	}
}
